plugins/fzuse.rb

```diff
@@ -0,0 +1,150 @@
+require 'socket'
+
+# this is the main routine that's executed in the grandchild process (msfconsole -> fzf -> this)
+if $PROGRAM_NAME == __FILE__
+  exit 64 unless ARGV.length == 2
+
+  UNIXSocket.open(ARGV[0]) do |sock|
+    sock.write ARGV[1] + "\n"
+    sock.flush
+
+    puts sock.read
+  end
+  exit 0
+end
+
+module Msf
+  class Plugin::FuzzyUse < Msf::Plugin
+    class ConsoleCommandDispatcher
+      include Msf::Ui::Console::CommandDispatcher
+
+      FZF_THEME = {
+        'fg' => '-1',
+        'fg+' => 'white:regular:bold',
+        'bg' => '-1',
+        'bg+' => '-1',
+        'hl' => '-1',
+        'hl+' => 'red:regular:bold',
+        'info' => '-1',
+        'marker' => '-1',
+        'prompt' => '-1',
+        'spinner' => '-1',
+        'pointer' => 'blue:bold',
+        'header' => '-1',
+        'border' => '-1',
+        'label' => '-1',
+        'query' => '-1'
+      }.freeze
+
+      def initialize(driver)
+        super
+
+        @module_dispatcher = Msf::Ui::Console::CommandDispatcher::Modules.new(driver)
+      end
+
+      def name
+        'FuzzyUse'
+      end
+
+      #
+      # Returns the hash of commands supported by this dispatcher.
+      #
+      def commands
+        {
+          'fzuse' => 'A fuzzy use command added by the FuzzyUse plugin'
+        }
+      end
+
+      def pipe_server(socket_path)
+        server = UNIXServer.new(socket_path)
+        File.chmod(0600, socket_path)
+        loop do
+          client = server.accept
+          begin
+            unless (input_string = client.gets&.chomp).blank?
+              if (mod = framework.modules.create(input_string))
+                client.puts(Serializer::ReadableText.dump_module(mod))
+              end
+            end
+          rescue StandardError
+          end
+          client.close
+        end
+      rescue EOFError
+      ensure
+        server.close if server
+        File.delete(socket_path) if File.exist?(socket_path)
+      end
+
+      #
+      # This method handles the fzuse command.
+      #
+      def cmd_fzuse(*args)
+        selection = nil
+
+        Dir.mktmpdir('msf-fzuse-') do |dir|
+          socket_path = File.join(dir, "msf-fzuse.sock")
```

Comment:
What are your thoughts on using Ruby's tmp file API and creating the file `~/.msf4/tmp/local` directory

```
>> Tempfile.create(['msf-fsuse-', '.sock'], Msf::Config.local_directory) { |socket_file| puts socket_file.path }
/home/vagrant/.msf4/local/msf-fsuse-20240927-2373170-2tm1kk.sock
```

Which I think should allow us to drop the wrapper `mktmpdir` call here too

Reply:
That doesn't work:

```
[6] pry(main)> Tempfile.create(['msf-fsuse-', '.sock']) { |sf| UNIXServer.new(sf) }
Errno::EADDRINUSE: Address already in use - connect(2) for /tmp/msf-fsuse-20240927-98954-b5sojv.sock
from (pry):6:in `initialize'
[7] pry(main)> Tempfile.create(['msf-fsuse-', '.sock']) { |sf| UNIXServer.new(sf.path + '.new') }
=> #<UNIXServer:/tmp/msf-fsuse-20240927-98954-of8yja.sock.new>
[8] pry(main)> 
```

Reply:
Ah that makes sense :+1:

So I guess this would be enough to move the file into the `~/.msf4` folder

```
>> Dir.mktmpdir(['msf-fzuse-', '.sock'], Msf::Config.local_directory) { |dir| socket_path = File.join(dir, "msf-fzuse.sock"); UNIXServer.new(socket_path) }
=> #<UNIXServer:/home/vagrant/.msf4/local/msf-fzuse-20240927-2373170-wckou3.sock/msf-fzuse.sock>
>> 
>> Dir.mktmpdir(['msf-fzuse-', '.sock'], Msf::Config.local_directory) { |dir| socket_path = File.join(dir, "msf-fzuse.sock"); UNIXServer.new(socket_path) }
=> #<UNIXServer:/home/vagrant/.msf4/local/msf-fzuse-20240927-2373170-jq5r0e.sock/msf-fzuse.sock>
>> 
```

Reply:
Is it necessary to put it in the config directory? The file only lasts a couple of seconds while the user is picking their module and I wasn't aware there was a precedent of placing temp files in the config directory.

Reply:
We've had PRs that put wordlists etc into places outside of `~/.msf4` which I think we've tried to avoid in the past, and for Metasploit Pro it's unexpected for files to appear in random places IMO instead of being self-contained. It's not a strong blocker for me :+1: